core/consensus/babe/babe_lottery.hpp

```diff
@@ -0,0 +1,64 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef KAGOME_BABE_LOTTERY_HPP
+#define KAGOME_BABE_LOTTERY_HPP
+
+#include <boost/optional.hpp>
+#include "consensus/babe/common.hpp"
+#include "crypto/crypto_types.hpp"
+
+namespace kagome::consensus {
+  /**
+   * Interface for acquiring leadership information for the current Babe epoch.
+   * It is expected to be used as follows:
+   *    - first epoch is started, r (randomness) = 0
+   *    - blocks are being accepted
+   *    - when a block is finalized, submitVRFValue function is called for each
+   *      block in the finalized chain up to the last finalized one (in the
+   *      chronological order). The chain can be retrieved via calling
+   *      BlockTree::longestPath() firstly and only then BlockTree::finalize(..)
+   *    - at the end of the epoch, computeRandomness(..) is called, providing a
+   *      randomness value for the new epoch
+   *    - that value is then is used in slotsLeadership(..)
+   *
+   */
+  struct BabeLottery {
+    virtual ~BabeLottery() = default;
+
+    /**
+     * Compute leadership for all slots in the given epoch
+     * @param epoch structure; "epoch_index" MUST be set to the epoch to compute
+     * randomness for
+     * @param keypair of this node
+     * @return vector of outputs; none means the peer was not chosen as a leader
+     * for that slot, value contains VRF value and proof
+     *
+     * @note must be called exactly ONCE per epoch, when it gets changed
```

Comment:
What would happen if it's called more than once and can it be prevented in the code?

Reply:
I afraid, it cannot be prevented (and shouldn't). The thing is, every time you call `submitVRFValue` in the same class, a new value is added to the accumulator. And this accumulator is nullified during the `slotsLeadership` call.
If that function is called twice per slot, it is a terrible mistake, which must not happen. I think, I will add a `BOOST_ASSERT` for that case to stop the program execution and signalize something went wrong.

Reply:
Also, that comment was for `computeRandomness`, not `slotsLeadership` :)